### ISSUE TITLE
Remove or fix broken commit 4bb41a1 "mitigate formspec exploits"

### ISSUE THREAD

**Issue by osjc** [minetest/minetest#8090]
The bug #7374: Formspec shown from another formspec is not being handled properly" is still present and is caused by a botched attempt to mitigate formspec exploits (commit 4bb41a1). The bug #7374 was closed because the simple testcase presented there is satisfied but there is a more complex testcase which still fails: Advanced Trains mod, the TSS branch (try to program a path).

The real bug here is in the protocol. When the client closes a form, it sends an empty formspec and that is the bug. It shall send the name of the formspec that was closed instead. The commit 4bb41a1 naively assumes that an empty string received as a form data submission signifies there is no form open in the client as in this example:

Server: "Here is formspec for [form1]"
Client: "I am displaying form [form1]"
Client: "I am submitting data from [form1]"
Server: "Here is formspec for [form2]" (responds to submission  of data from [form1])
Client: "" (trying to say "[form1] was closed")
Client: "I am displaying data from [form2]"
Client: "I am submitting data from [form2]"
Client: "" (trying to say "[form2] was closed")

The problem is that UDP is not a reliable protocol and as such the message with the "empty string as form submission" could get reordered past another message containing a new formspec to be opened:

Server: "Here is formspec for [form1]"
Client: "I am displaying form [form1]"
Client: "I am submitting data from [form1]"
Server: "Here is formspec for [form2]" (responds to submission  of data from [form1])
Client: "I am displaying data from [form2]"
Client: "" (trying to say "[form1] was closed")
Client: "I am submitting data from [form2]"
Client: "" (trying to say "[form2] was closed")

or even:

Server: "Here is formspec for [form1]"
Client: "I am displaying form [form1]"
Client: "I am submitting data from [form1]"
Client: "" (trying to say "[form1] was closed")
Server: "Here is formspec for [form2]" (responds to submission  of data from [form1])
Client: "I am displaying data from [form2]"
Client: "" (trying to say "[form2] was closed")
Client: "I am submitting data from [form2]"

In both cases the mitigation will think [form2] is closed even when it is not and thus exhibit the broken behavior described in #7374 (yet again :( ).

As a matter of fact I added some data dumping commands in the relevant places (see branchhttps://github.com/osjc/minetest/tree/TryingToTrackDownFormspecs, commit 3f470f7 (the top one))

2019-01-10 23:11:23: ACTION[Server]: Client 2 passed form at_il_rprog_(3,23,-4)
2019-01-10 23:11:26: ACTION[Server]: Client 2 submitted data for form: at_il_rprog_(3,23,-4)
2019-01-10 23:11:34: ACTION[Server]: Client 2 passed form data: size[7,6]label[0.5,0.5;Finish programming route]label[0.5,1.5;Route ends at signal:]label[0.5,2  ;Signal at (3,23,-4)]field[0.8,3.5;5.2,1;name;Enter Route Name;]button_exit[0.5,4.5;  5,1;save;Save Route]
2019-01-10 23:11:34: ACTION[Server]: Client 2 passed form at_il_routepf
2019-01-10 23:11:44: ACTION[Server]: Client 2 passed form data: 
2019-01-10 23:11:44: ACTION[Server]: Client 2 passed an empty form.
2019-01-10 23:11:48: ACTION[Server]: Client 2 submitted data for form: at_il_routepf
2019-01-10 23:11:48: ACTION[Server]: 'singleplayer' submitted formspec ('at_il_routepf') but server hasn't sent formspec to client, possible exploitation attempt

Here [form1] is "at_il_rprog_(3,23,-4)" and [form2] is "format_il_routepf". Note the "Client 2 passed form data:" and "Client 2 passed an empty form." messages before the "Client 2 submitted data for form: at_il_routepf" but after the "Client 2 passed form at_il_routepf". These two lines represent the reordered "Client closed [form1]" from the examples above.

For the "at_il_rprog_(3,23,-4)" and [form2] is "format_il_routepf" forms see the TSS branch download at https://forum.minetest.net/viewtopic.php?f=11&t=14726&sid=47b154987311ccd39869647e8b2bfb1f&start=1125#p339307.

Digging further with the debugger revealed that the offending lines are produced by Server::handleCommand_InventoryFields and as such are caused by the client's asynchronous message about the previous closed form.

EDIT: I realized that the "Client 2 passed form XXX" shall actually be "Server passed formspec XXX to Client 2". Formspecs originate at the server.

**Comment by osjc** [comment]
[testcase.tar.gz](https://github.com/minetest/minetest/files/2748132/testcase.tar.gz)
World with the TSS branch exhibiting the bug. This was used to catch it and produce the debug.txt whose part is shown in the report. Use:

mv testcase.tar.gz testcase.tar.xz # bug GitHub about support for tar.xz files
mkdir workdir
cd workdir
tar xJvf testcase.tar.xz
git clone <url-of-minetest-project>
git clone <url-of-minetest-game-project>
sh testbuild.sh # go make some tea ;-)
sh testrun.sh

Now in the window that shows up open the "BugTest" world. Turn until you see a signal facing AWAY from you, then right-click it and click "New Route". Now turn around and punch the little box next to the following signal. A form shall open with a button "Finish the route HERE", click that button. Write something into the text box that appears and click "save". If the code works, there shall be no markers on the rail.

The "dirt block" between the two signals is the small testcase from #7374. It works with the improper fix.

**Comment by osjc** [comment]
So, a hotfix that removed the fix which turned out to be worse than the disease it tried to address was just submitted. The following things now shall be implemented:

1. Protocol change: Once the client receives a formspec to be displayed and when the time arrives to display it, it checks that there is a name on it and reports that name back to the server in a "form displayed" message. Stuff like the inventory form shall have autogenerated names like "inventory_of_player_xyz" for the player's inventory or "player_xyz_accessing_node_a_b_c" for player right-clicking a node with an unnamed formspec attached to it. Then the client shows the formspec to the server. When the player clicks a button or does something else that causes the client to send the fields to the server, a "fields received, form still open" message (which already contains the name of the form) shall be produced as usual. If the player cancels the form (ESC) or the form "cancels itself" (some buttons may instruct the client to close the form and discard any data in it), the client shall send the name of the form again in a message "form cancelled". And, finally if the player uses a button like "OK", the message "fields received, form closed" shall be produced (do NOT split this into two messages as they may get messed up.
2. Improve the formspecs to allow creating buttons that send the data from the form and close it in one step.
3. Now the server has all the data it needs to mitigate the "forged formspec" exploit attempts. The server needs to track which forms is/are open and what buttons are there. Then it can make sure that if for example "OK"-type button was specified in the form data, the form is considered closed after processing that form data (so a further attempt to submit such data will be blocked by the security check)

**Comment by rubenwardy** [comment]
There's some inaccuraries here:

1. Formspecs should be sent over Minetest's reliability layer. If they're not then there's a bug here
2. The client tells the server that a form has closed like so:

   ```lua
   {
       formname = "mymod:formname",
       fields = { quit = true },
   }
   ```

   So the server is well aware.
3. The client never tells the server that it has opened a formspec

**Comment by osjc** [comment]
A description of an idea about a hotfix that does not get rid of the broken commit completely was added to #8091.

**Comment by osjc** [comment]
I left my optimization work alone for now because I started to get ideas how to fix this. See the pull request #8091 for "public fix development".

**Comment by osjc** [comment]
As @rubenwardy pointed out, it turned out the protocol already has the data we need to do proper form security.

**Comment by osjc** [comment]
It turns out the "form1" from the "form1 closed" didn't get lost anywhere. The security code merely didn't bother to take a look at it before taking action. It turns out I "had my hand" on the form name in that case all the time. The function in question gets it. The only bug that was there was that when the form was closed ("formspec.empty"), the security code didn't bother checking that the form being closed is indeed the one that is currently recorded as open.

**Comment by osjc** [comment]
> Well, not so well aware when the form security part only sees an empty string in that case.

Turns out the form security part sees only empty string because it doesn't bother to look for the real thing. But fixed it. Hopefully.

Funny how I answered my own question.

**Comment by jastevenson303** [comment]
`2019-01-10 23:11:48: ACTION[Server]: 'singleplayer' submitted formspec ('at_il_routepf') but server hasn't sent formspec to client, possible exploitation attempt`

I had gotten this error a lot, and was able to remove it by reworking the formspec code.  minetest.show_formspec closes any already open formspec.

Are you sure it's not just the order of things?  (Sorry for my dumb-ness, please excuse me.)

**Comment by rubenwardy** [comment]
Is this still an issue with the PR merged?
